evidence/reactor.go

```diff
@@ -169,6 +169,11 @@ func (evR Reactor) checkSendEvidenceMessage(
 		return nil, true
 	}
 
+	if evR.evpool.Size() == 0 {
+		// XXX: We have evidence in the clist which no longer exists in the pool.
+		return nil, false
+	}
```

Comment:
This logic seems odd. Can this be vulnerable to race conditions and such? I.e. if it's been removed from the pool for a valid reason but not actually sent yet?

Reply:
Yeah this is my debatable hack. If it gets removed from the evidence pool it should always also be removed from the clist

Reply:
The other problem is this hack doesn't work if there is other evidence

Reply:
I don't mind removing it. It just means that we will send occasionally a harmless message about an evidence that is already committed. Not the worst thing to happen.

Reply:
Yeah, I suspect this might be vulnerable to race conditions, and the question then is what's worse: sending a harmless message, or not sending evidence at all?

I'd be interested to get to the bottom of this and find the root problem though. Wouldn't surprise me if that turns out to be the same race condition, i.e. that the clist and evidence pool aren't both updated atomically in a single critical section.

Reply:
yes it could be possible that one of the go routines broadcasting evidence to a peer reads the evidence and prepares to send it just before the evidence pool's write operation (or in this case a delete). But the fact that this occurs a couple of heights after and only to this full node who is a persistent peer adds a different dimension

Reply:
Ahhh I think I've got it